### PR TITLE
add source publisher

### PR DESCRIPTION
**Pull request by adrianschroeter**

### CODE REVIEW

File: xml/obs_ag_source_publish.xml [hunk 1]
```diff
@@ -0,0 +1,42 @@
+<!DOCTYPE sect1
+[
+  <!ENTITY % entities SYSTEM "entity-decl.ent">
+    %entities;
+]>
+<sect1 xmlns="http://docbook.org/ns/docbook"
+ xmlns:xi="http://www.w3.org/2001/XInclude"
+ xmlns:xlink="http://www.w3.org/1999/xlink" version="5.1"
+ xml:id="_source_publish">
```
Comment (sknorr) [inline xml/obs_ag_source_publish.xml:9]: Why use underscores at the beginning of IDs (repeated also for the other IDs in this file)? (Or are those auto-generated via AsciiDoctor?)

I'd suggest to just remove those.

File: xml/obs_admin_components.xml [hunk 2]
```diff
@@ -76,6 +76,14 @@
     directory on the back-end. Each OBS installation has one publisher per
     partition. </para>
   </sect3>
+  <sect3>
+   <title>Source Publisher</title>
+   <para> The source publisher processes "sourcepublish" events from the publisher for
+    finished repositories. It needs to run on the same server as the source server.
```
Comment (sknorr) [inline xml/obs_admin_components.xml:82]: * "finished repository" sounds wrong. Also not 100% sure what is meant -- "a repository in which all jobs have finished building"? 
* I guess this is a bigger terminology issue -- but using "server" for two different things (software component v/ machine) in the same sentence is a bit confusing.

File: xml/obs_admin_components.xml [hunk 3]
```diff
@@ -76,6 +76,14 @@
...
+    It can be used to publish a filesystem structure providing all sources of
+    published binaries. In case of images or containers this includes also the sources
```
Comment (sknorr) [inline xml/obs_admin_components.xml:84]: style improvement: `includes also` -> `also includes`

File: xml/obs_ag_source_publish.xml [hunk 4]
```diff
@@ -0,0 +1,42 @@
...
+ <title>Source Publisher</title>
+ <para>The job of the source publish service is to publish all sources for
```
Comment (sknorr) [inline xml/obs_ag_source_publish.xml:11]: * `source-publishing service` would sound much better to me, grammar-wise.
* You can make the sentence shorter by replacing `The job of the source publish service is to publish ...` with `The source-publishing service publishes ...`, but even then it sounds a little redundant. ;/

File: xml/obs_ag_source_publish.xml [hunk 5]
```diff
@@ -0,0 +1,42 @@
...
+directly before published binaries. This will also the sources of repackaged
+binaries. For example the sources of rpms which are used inside of product, 
+appliance or container images. A prerequisite for this is that OBS and the
+build tooling supports content tracking.</para>
+
+ <sect2 xml:id="_source_publish_configuration">
+  <title>Configuring Source Publisher</title>
+  <para>The source publishing can be configured via the file
+    <emphasis>/usr/lib/obs/server/BSConfig.pm</emphasis>, where it can be 
+    global enabled or just for some projects. It is possible to use regular
```
Comment (sknorr) [inline xml/obs_ag_source_publish.xml:21]: grammar: `global enabled` -> `enabled globally`

File: xml/obs_ag_source_publish.xml [hunk 6]
```diff
@@ -0,0 +1,42 @@
...
+   expressions here.</para>
+  <para>The publishing can be enabled by defining the rsync module to push the content:</para>
+  <screen language="perl">our $sourcepublish_sync = 'rsync://your_rsync_server/rsync_module'; </screen>
+  <para>By default every project get published, but it is possible to define a whitelist via:</para>
+  <screen language="perl">our $sourcepublish_filter = [ "openSUSE:.*", "SUSE:.*" ]; </screen>
+ </sect2>
+
+ <sect2 xml:id="_source_publish_considerations">
+  <title>Considerations</title>
+  <para>The source publishing is publishing the sources as they are hosted in &obs;. This
```
Comment (sknorr) [inline xml/obs_ag_source_publish.xml:31]: Do you need the word "service" here? `The source-publishing service is ...`

Comment (sknorr) [inline xml/obs_ag_source_publish.xml:31]: Do you need the word "service" here? `The source-publishing service is ...`

File: xml/obs_ag_source_publish.xml [hunk 7]
```diff
@@ -0,0 +1,42 @@
...
+        means these are the unprocessed sources and the content is not identical to the
+        content of source rpms for example. Instead these are the sources which are the
+        base for source rpms.</para>
```
Comment (sknorr) [inline xml/obs_ag_source_publish.xml:34]: You're using "source" a few too many time within the span of a single para :) I've tried to shorten it a bit, and I hope to have kept all the (necessary) content:


```suggestion
  <para>
   The source-publishing service publishes unprocessed sources, exactly the way they are
   in &obs;.
   In contrast, source RPMs may instead use processed sources that differ from those in
   &obs; (but are derived from them).
  </para>
```

File: xml/obs_ag_source_publish.xml [hunk 8]
```diff
@@ -0,0 +1,42 @@
...
+  <para>As a consequence hints like NoSource: tags in spec files are ignored. The only
+        way to disable publishing for the user is to disable access or sourceaccess via
+        the flags.</para>
+  <para>The filesystem structure is <emphasis>$project/$package/$srcmd5/</emphasis>. A &disturl;
+        inside of binary builds can be used to find the right sources.</para>
+  <para>&obs; will care for de-duplication on the rsync server. This must get implemented there.</para>
```
Comment (sknorr) [inline xml/obs_ag_source_publish.xml:40]: * wording correction: `&obs; will take care of deduplication ...`
* I am a bit confused by the final sentence -- is this feature already implemented? If so, I'd just remove that sentence.

**Comment by sknorr** [comment]
Hey, are you still working on this?

File: xml/obs_ag_source_publish.xml (hunk 8 above)
Comment (adrianschroeter) [reply xml/obs_ag_source_publish.xml:40]: the rsync server is not part of OBS. this is something admin has to take care of.

**Comment by sknorr** [comment]
Last (force-)push to this branch seems like an accident. (I can probably restore from my non-updated repo if you like.)

**Comment by adrianschroeter** [comment]
@sknorr urgs, yes, please

will work on the fixes directly afterwards, I promise ...

**Comment by adrianschroeter** [comment]
@sknorr found it in my reflog ...

File: xml/obs_ag_source_publish.xml [hunk 9]
```diff
@@ -0,0 +1,42 @@
...
+ <sect2 xml:id="_source_publish_considerations">
+  <title>Considerations</title>
+  <para>The source publishing is publishing the sources as they are hosted in &obs;. This
+        means these are the unprocessed sources and the content is not identical to the
+        content of source rpms for example. Instead these are the sources which are the
+        base for source rpms.</para>
+  <para>As a consequence hints like NoSource: tags in spec files are ignored. The only
+        way to disable publishing for the user is to disable access or sourceaccess via
+        the flags.</para>
+  <para>The filesystem structure is <emphasis>$project/$package/$srcmd5/</emphasis>. A &disturl;
```
[the remark this answers is not in the document]
Comment (adrianschroeter) [reply xml/obs_ag_source_publish.xml:38]: added

File: xml/book-obs-admin-guide.xml [hunk 10]
```diff
@@ -36,6 +36,7 @@
  <xi:include href="obs_ag_installation_and_configuration.xml"/>
  <xi:include href="obs_ag_overview_filesystem.xml"/>
  <xi:include href="obs_ag_administration.xml"/>
+ <xi:include href="obs_ag_source_publish"/>
```
[the remark this answers is not in the document]
Comment (adrianschroeter) [reply xml/book-obs-admin-guide.xml:39]: fixed

File: xml/obs_ag_source_publish.xml (hunk 1 above)
Comment (adrianschroeter) [reply xml/obs_ag_source_publish.xml:9]: IIRC copy and paste ... changed

File: xml/obs_admin_components.xml (hunk 2 above)
Comment (adrianschroeter) [reply xml/obs_admin_components.xml:82]: using "instance" now

File: xml/obs_admin_components.xml (hunk 3 above)
Comment (adrianschroeter) [reply xml/obs_admin_components.xml:84]: changed

File: xml/obs_ag_source_publish.xml [hunk 11]
```diff
@@ -0,0 +1,42 @@
...
+appliance or container images. A prerequisite for this is that OBS and the
+build tooling supports content tracking.</para>
+
+ <sect2 xml:id="_source_publish_configuration">
+  <title>Configuring Source Publisher</title>
+  <para>The source publishing can be configured via the file
+    <emphasis>/usr/lib/obs/server/BSConfig.pm</emphasis>, where it can be 
+    global enabled or just for some projects. It is possible to use regular
+   expressions here.</para>
+  <para>The publishing can be enabled by defining the rsync module to push the content:</para>
```
[the remark this answers is not in the document]
Comment (adrianschroeter) [reply xml/obs_ag_source_publish.xml:23]: changed. rsync is naming its targets module, so I followed that

File: xml/obs_ag_source_publish.xml (hunk 6 above)
Comment (adrianschroeter) [reply xml/obs_ag_source_publish.xml:31]: k

**Comment by adrianschroeter** [comment]
I tried to add the disturl definition, but failed to get the right syntax, can you help me?

File: xml/obs_glossary.xml [hunk 12]
```diff
@@ -329,6 +329,18 @@
   <glossterm>Diff</glossterm>
   <glosssee otherterm="obs.glos.patch"/>
  </glossentry>
+ <glossentry xml:id="obs.glos.disturl">
+  <glossterm>DISTURL</glossterm>
+  <glossdef>
+   <para>
+    The DISTURL is a unique identifier of a source and it's build setup. It is written
+    usually written inside of the build result to be able to identify the origin.
+    A DISTURL is structured as <command>obs://OBS_NAME/PROJECT/REPOSITORY/REVISION-PACKAGE</command>.
```
Comment (tomschr) [inline xml/obs_glossary.xml:338]: The content in `<command>` is actually an URI. As such, I'd suggest:

```suggestion
    A DISTURL is structured as <uri>obs://OBS_NAME/PROJECT/REPOSITORY/REVISION-PACKAGE</uri>.
```

**Comment by tomschr** [comment]
> I tried to add the disturl definition, but failed to get the right syntax, can you help me?

The general syntax would be:

```xml
<!ENTITY disturl "...">
```

However, to include a link makes it a little bit more complicated. You need to add two namespaces (DocBook5 and XLink), otherwise XML complains.  If you want to add a link for `disturl` entity, insert the following line inside `entity-decl.ent`:

```xml
<!ENTITY disturl "<link xlink:href='http://YOUR_URL_GOES_HERE'
  xmlns='http://docbook.org/ns/docbook'
  xmlns:xlink='http://www.w3.org/1999/xlink' />
">
```

Hope this helps.

**Comment by adrianschroeter** [comment]
@tomschr I think a link to the glossary reference would be better. I do not have any external link.

**Comment by tomschr** [comment]
> @tomschr I think a link to the glossary reference would be better. I do not have any external link.

@adrianschroeter Ahh, I see. If you want just an internal reference to the glossary, use `<xref linkend="obs.glossary"/>`. You can reference a specific glossary item too.

**Comment by adrianschroeter** [comment]
okay, fixed that, but I do not get the meaning of the validation error...

**Comment by tomschr** [comment]
> okay, fixed that, but I do not get the meaning of the validation error...

The file `obs_ag_source_publish.xml` contains a root element `sect1`. This file is included in `xml/book-obs-admin-guide.xml`. However, that is a `book`, so you end up with this structure:

```
book
   title
   sect1
```

This is not allowed. A `sect1` can only appear inside a `chapter`, `appendix` and a few others. You can solve this be either one of the following solutions:

1. Find a chapter where this could belong and include it there.
2. Change the root element from `sect1` to `chapter`.

Hope this helps.

**Comment by adrianschroeter** [comment]
IMHO everything is fine now

**Comment by sknorr** [comment]
There are a few comments unhelpfully hidden by GitHub that should probably still be resolved. At the very least this one: https://github.com/openSUSE/obs-docu/pull/149/files#r434690328

File: xml/obs_ag_source_publish.xml [hunk 13]
```diff
@@ -0,0 +1,42 @@
...
+<sect1 xmlns="http://docbook.org/ns/docbook"
+ xmlns:xi="http://www.w3.org/2001/XInclude"
+ xmlns:xlink="http://www.w3.org/1999/xlink" version="5.1"
+ xml:id="_source_publish">
+ <title>Source Publisher</title>
+ <para>The job of the source publish service is to publish all sources for 
+directly before published binaries. This will also the sources of repackaged
+binaries. For example the sources of rpms which are used inside of product, 
+appliance or container images. A prerequisite for this is that OBS and the
+build tooling supports content tracking.</para>
```
[the remark this answers is not in the document]
Comment (adrianschroeter) [reply xml/obs_ag_source_publish.xml:15]: it is optional and can be configured per project. Will try to simplify this with next push

File: xml/obs_ag_source_publish.xml (hunk 5 above)
Comment (adrianschroeter) [reply xml/obs_ag_source_publish.xml:21]: okay

**Pull request merged**